Redesigning `protostar.toml` 4 — Saving `ConfigurationFileV2`

kasperski95 opened the pull request.

mkaput commented on protostar/configuration_file/confiuration_toml_writer_test.py:
```diff
@@ -0,0 +1,54 @@
+        filepath=output_file_path,
+    )
+
+    result = output_file_path.read_text()
+    assert result is not None
+    assert not '["' in result
+    assert not '"]' in result
```
wouldn't it be more fault-proof to verify whole file contents exactly?

mkaput commented on protostar/configuration_file/confiuration_toml_writer_test.py:
```diff
@@ -0,0 +1,54 @@
+from dataclasses import dataclass
```
I'm missing a test that verifies that `ConfigurationTOMLWriter` writes stuff in fixed order. So that our writes will not cause nondeterministic diffs.

kasperski95 replied:
`test_saving_configuration` is a snapshot test, and it doesn't fail randomly

mkaput commented on protostar/configuration_file/configuration_file_v2.py:
```diff
@@ -89,5 +96,48 @@ def read(
 
-    def save(self, configuration_file_model: ConfigurationFileV2Model) -> Path:
-        raise NotImplementedError("Operation not supported.")
+    def create_file_content(
```
I wonder if it wouldn't be simpler to simply build a Python dict containing all data and pass it to the writer.

In general that would drop the need for the `ConfigurationFileContentBuilder` class.

mkaput replied:
Shouldn't `profile` en/decoding be shifted to a separate class that would be layered over all of this? You're making too powerful objects here.

kasperski95 replied:
- there might be different conventions for encoding `profiles` per format 
- `ConfigurationTOMLContentBuilder` has 54 lines of cohesive code
- there are more important things to do

mkaput reviewed (changes requested):
It's not clear to me how to write something like this, with exactly this formatting:

```toml
[dependencies]
foo = "^1.0.0"

[profile.foo.dependencies]
foo = { git = "...", branch = "..." }
```

I know that this is out of scope of this PR because we do not have such feature, but I don't know how to fit it into this design.

kasperski95 commented:
> It's not clear to me how to write something like this, with exactly this formatting:

Added a test case that represents how to do it. I noticed inline tables weren't supported, so I replaced the `toml` package which is going to be useful anyway to add comments.

mkaput commented on protostar/configuration_file/configuration_file_v2_test.py:
```diff
@@ -144,3 +138,31 @@ def test_saving_configuration(
     assert result == protostar_toml_content
+
+
+def test_saving_in_particular_order(
```
I think 1. building the configuration file ad hoc specifically for this test and 2. testing exact contents would be more robust here and 3. testing only ConfigurationTOMLWriter independent of the model.

kasperski95 replied:
> 3. testing only ConfigurationTOMLWriter independent of the model.

I'll remove `ConfigurationTOMLWriter`. `ConfigurationTOMLContentBuilder` is what should be tested.

kasperski95 commented on pyproject.toml:
```diff
@@ -26,6 +26,7 @@ tomli = "<2.0.0"
 tomli-w = "^1.0.0"
 tqdm = "^4.64.0"
 typing-extensions = "^4.0.1"
+tomlkit = "^0.11.4"
```
Maybe in the future. I'd leave the old library as long as we support V1.

kasperski95 commented on poetry.lock:
```diff
@@ -54,7 +54,7 @@ optional = false
 
 [package.extras]
-tests = ["mypy (>=0.800)", "pytest", "pytest-asyncio"]
+tests = ["pytest", "pytest-asyncio", "mypy (>=0.800)"]
```
I believe the most recent one (I need to check it on my second laptop).

kasperski95 replied:
Reverted changes

kasperski95 commented on protostar/configuration_file/configuration_toml_content_builder.py:
```diff
@@ -0,0 +1,64 @@
+
+
+class ConfigurationTOMLContentBuilder(
+    ConfigurationFileContentBuilder[ConfigurationTOMLContent]
```
It's not necessary. It's a leftover after a different flow.

kasperski95 replied:
Removed generic

kasperski95 commented on protostar/configuration_file/configuration_toml_content_builder_test.py:
```diff
@@ -0,0 +1,97 @@
+        model=model,
+    )
+
+    assert "[profile.devnet.declare]" in result
```
Done

mkaput reviewed (approved):
LGTM
